src/c_ffi.rs

```diff
@@ -0,0 +1,34 @@
+use libc::{c_char,size_t};
+use std::ffi::CStr;
+use std::vec;
+
+use super::Trie;
+use super::trie_common::TrieCommon;
+
+ffi_fn! {
+    fn radix_trie_create()-> *const Trie<Vec<u8>, usize>{
+        let trie = Trie::<Vec<u8>, usize>::new();
+        return Box::into_raw(Box::new(trie));
+    }
+}
+
+ffi_fn! {
+    fn radix_trie_free(trie_ptr: *const Trie<Vec<u8>, usize>){
+        unsafe { Box::from_raw(trie_ptr as *mut Trie<Vec<u8>, usize>); }
+    }
+}
+
+ffi_fn! {
+    fn radix_trie_insert(trie_ptr:*const Trie<Vec<u8>, usize>, key_ptr:*const c_char, value:usize){
+        let mut trie = unsafe { &mut *(trie_ptr as *mut Trie<Vec<u8>, usize>) };
+        let key =  unsafe { CStr::from_ptr(key_ptr) }.to_bytes().to_vec();
```

Comment:
We could optimise this (avoid a double allocation), by implementing `TrieKey` for `CStr`. Using a `Vec<u8>` as the key means that it gets cloned each time (see: https://github.com/michaelsproul/rust_radix_trie/blob/c34a167f424db35583612b61051d8441e35f5f2a/src/keys.rs#L85-L89)

Reply:
Actually, I did this on purpose to avoid expected string pointer deallocation from foreign languages, especially managed ones whose memory is not being handled explicitly, does that make sense?

Reply:
Yeah, I agree that we want that behaviour, but the current code copies the string twice. The `.to_bytes().to_vec()` call will copy everything into a `Vec<u8>`, and when that `Vec<u8>` gets inserted in the trie it will get cloned. Instead, if we generate the `Vec<u8>` in the `TrieKey::encode_bytes` function, the string will only get copied once, like:

```rust
impl TrieKey for CStr {
    fn encode_bytes(&self) -> Vec<u8> {
        self.to_bytes().to_vec()
    }
}
```

Maybe the compiler is smart enough to optimise this away, but it doesn't hurt to give it some help (and `impl TrieKey for CStr` is generally useful anyway)